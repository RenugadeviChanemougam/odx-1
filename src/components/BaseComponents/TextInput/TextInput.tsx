import React from 'react';
import PropTypes from 'prop-types';
import FormGroup from '../FormGroup/FormGroup';

export default function TextInput(props){

  const {name, errorText, hintText, inputProps} = props;

  const inputClasses = `govuk-input ${errorText && 'govuk-input--error'}`

  //TODO - Handle input widths
  //TODO - Handle input types (password, email, numeric) - Or investigate if these should be separate components, or can simple be handled by inputProps
  //TODO - Handle autocomplete settings

  // TODO Investigate if this approach can be streamlined and/or refactored for use in other components
  const hintID = `${name}-hint`;
  const errorID = `${name}-error`;
  const describedByIDs = [hintText && hintID, errorText && errorID];

  return(
    <FormGroup {...props}>
      <input className={inputClasses} {...inputProps} id={name} name={name} aria-describedby={describedByIDs.join(' ').trim()}></input>
    </FormGroup>
    )
}

TextInput.propTypes = {
  ...FormGroup.propTypes,
  name: PropTypes.string,
  inputProps: PropTypes.object,
}
