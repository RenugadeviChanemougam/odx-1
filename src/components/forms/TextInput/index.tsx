import React, { useEffect, useState } from 'react';
import GDSTextInput from '../../BaseComponents/TextInput/TextInput';
import useAddErrorToPageTitle from '../../../helpers/hooks/useAddErrorToPageTitle';

declare const PCore;

export default function TextInput(props) {
  const {
    label,
    value = '',
    validatemessage,
    onChange,
    helperText,
    getPConnect,
    inputProps,
  } = props;

  // const maxLength = fieldMetadata?.maxLength;



  const [isOnlyQuestion, setIsOnlyQuestion] = useState(PCore.getFormUtils().getEditableFields("root/primary_1/workarea_1").length === 1);
  const [displayLabel, setDisplayLabel] = useState(label);
  useAddErrorToPageTitle(validatemessage);

  /* let testProp = {};
  testProp = {
    'data-test-id': testId
  }; */

  useEffect ( () => {
    setIsOnlyQuestion(PCore.getFormUtils().getEditableFields("root/primary_1/workarea_1").length === 1);
  }, [])

  useEffect ( () => {
    if(isOnlyQuestion){
      setDisplayLabel(getPConnect().getDataObject()?.caseInfo.assignments[0].name);
    } else {
      setDisplayLabel(label);
    }
  }, [isOnlyQuestion])

  const extraInputProps = {onChange, value};

  // TODO Investigate more robust way to check if we should display as password
  if(label === "Password"){
    extraInputProps["type"]="password";
  }

  return (
    <>
    <GDSTextInput
      inputProps={{
        ...inputProps,
        ...extraInputProps
      }}
      hintText={helperText}
      errorText={validatemessage}
      label={displayLabel}
      labelIsHeading={isOnlyQuestion}
    />
    </>
  );
}
