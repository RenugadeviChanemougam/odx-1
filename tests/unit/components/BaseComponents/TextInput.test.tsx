import React from 'react';
import renderer from 'react-test-renderer';
import TextInput from '../../../../src/components/BaseComponents/TextInput/TextInput';

it('renders correctly', () => {
  const tree = renderer
    .create(<TextInput>Facebook</TextInput>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

let name = "name";
let label = "label";

it('Basic single question text input', () => {
  let a = TextInput.propTypes;

  const tree = renderer
    .create(<TextInput name={name} label={label} labelIsHeading={true}></TextInput>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Basic single question password input', () => {
  let a = TextInput.propTypes;

  const tree = renderer
    .create(<TextInput name={name} label={label} labelIsHeading={true} inputProps={{type:"password"}}></TextInput>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Basic multi question text input', () => {
  let a = TextInput.propTypes;

  const tree = renderer
    .create(<TextInput name={name} label={label} labelIsHeading={false}></TextInput>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});



it('Basic multi question text input with error', () => {
  let a = TextInput.propTypes;

  const tree = renderer
    .create(<TextInput name={name} label={label} labelIsHeading={false} errorText={"Something went wrong"}></TextInput>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});




it('Basic multi question text input with hint', () => {
  let a = TextInput.propTypes;

  const tree = renderer
    .create(<TextInput name={name} label={label} labelIsHeading={false} hintText={"Enter some text"}></TextInput>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});



it('Basic multi question text input with hint and error', () => {
  let a = TextInput.propTypes;

  const tree = renderer
    .create(<TextInput name={name} label={label} labelIsHeading={false} hintText={"Enter some text"} errorText={"Something went wrong"}></TextInput>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
