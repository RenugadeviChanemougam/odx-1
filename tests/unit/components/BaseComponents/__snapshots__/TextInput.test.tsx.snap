// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic multi question text input 1`] = `
<div
  className="govuk-form-group"
>
  <label
    className="govuk-label"
    htmlFor="name"
  >
    label
  </label>
  <input
    aria-describedby=""
    className="govuk-input govuk-input--width-undefined"
    name="name"
  />
</div>
`;

exports[`Basic multi question text input with error 1`] = `
<div
  className="govuk-form-group govuk-form-group--error"
>
  <label
    className="govuk-label"
    htmlFor="name"
  >
    label
  </label>
  <p
    className="govuk-error-message"
    id="name-error"
  >
    <span
      className="govuk-visually-hidden"
    >
      Error:
    </span>
    Something went wrong
  </p>
  <input
    aria-describedby="name-error"
    className="govuk-input govuk-input--error govuk-input--width-undefined"
    name="name"
  />
</div>
`;

exports[`Basic multi question text input with hint 1`] = `
<div
  className="govuk-form-group"
>
  <label
    className="govuk-label"
    htmlFor="name"
  >
    label
  </label>
  <div
    className="govuk-hint"
    id="name-hint"
  >
    Enter some text
  </div>
  <input
    aria-describedby="name-hint"
    className="govuk-input govuk-input--width-undefined"
    name="name"
  />
</div>
`;

exports[`Basic multi question text input with hint and error 1`] = `
<div
  className="govuk-form-group govuk-form-group--error"
>
  <label
    className="govuk-label"
    htmlFor="name"
  >
    label
  </label>
  <div
    className="govuk-hint"
    id="name-hint"
  >
    Enter some text
  </div>
  <p
    className="govuk-error-message"
    id="name-error"
  >
    <span
      className="govuk-visually-hidden"
    >
      Error:
    </span>
    Something went wrong
  </p>
  <input
    aria-describedby="name-error name-hint"
    className="govuk-input govuk-input--error govuk-input--width-undefined"
    name="name"
  />
</div>
`;

exports[`Basic single question password input 1`] = `
<div
  className="govuk-form-group"
>
  <h1
    className="govuk-label-wrapper"
  >
    <label
      className="govuk-label govuk-label--l"
      htmlFor="name"
    >
      label
    </label>
  </h1>
  <input
    aria-describedby=""
    className="govuk-input govuk-input--width-undefined"
    name="name"
    type="password"
  />
</div>
`;

exports[`Basic single question text input 1`] = `
<div
  className="govuk-form-group"
>
  <h1
    className="govuk-label-wrapper"
  >
    <label
      className="govuk-label govuk-label--l"
      htmlFor="name"
    >
      label
    </label>
  </h1>
  <input
    aria-describedby=""
    className="govuk-input govuk-input--width-undefined"
    name="name"
  />
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="govuk-form-group"
>
  <h1
    className="govuk-label-wrapper"
  >
    <label
      className="govuk-label govuk-label--l"
    />
  </h1>
  <input
    aria-describedby=""
    className="govuk-input govuk-input--width-undefined"
  />
</div>
`;
